### PR TITLE
[feature] Interaction requests client api + settings panel

### PR DESCRIPTION
**Pull request by tsmethurst**
# Description

> If this is a code change, please include a summary of what you've coded, and link to the issue(s) it closes/implements.
>
> If this is a documentation change, please briefly describe what you've changed and why.

This pull request adds review and rejection / approval of interaction requests via the client API and settings panel.

Screenshot from 2024-08-19 14-32-41

Screenshot from 2024-08-19 14-32-51

To support this, requests and rejections are now stored in the database in addition to approvals, as that was just the easiest way of making it possible to page through requests (and, in future, page through approvals and rejections).

The structure of the approvals table has also changed. To make things a bit easier for myself, the current table is just dropped and recreated with the new schema, which means existing approvals will be lost. Since approvals were not in a release yet, I don't think this is a big deal, since only people running snapshots had interaction policies up to now (and therefore stored Accepts), but I welcome feedback on this and can make more of an effort to preserve existing Accepts if we decide it's necessary to do this.

~~Still draft because I'd still like to do the following:~~

- ~~[ ] Clean up any lingering unique indexes from dropped table (if db type = postgres)~~ Not necessary.
- [x] Add some more tests.
- ~~[ ] Have the interaction filter check if a previous InteractionRejection exists for a fave or boost of a status by a given account, and automatically reject reattempts (so people can't spam interaction requests at you).~~ I'll do this in a separate PR + it might not be necessary or desirable anyway, not sure yet.

Part of #3057 

## Checklist

Please put an x inside each checkbox to indicate that you've read and followed it: `[ ]` -> `[x]`

If this is a documentation change, only the first checkbox must be filled (you can delete the others if you want).

- [x] I/we have read the [GoToSocial contribution guidelines](https://github.com/superseriousbusiness/gotosocial/blob/main/CONTRIBUTING.md).
- [x] I/we have discussed the proposed changes already, either in an issue on the repository, or in the Matrix chat.
- [x] I/we have not leveraged AI to create the proposed changes.
- [x] I/we have performed a self-review of added code.
- [x] I/we have written code that is legible and maintainable by others.
- [x] I/we have commented the added code, particularly in hard-to-understand areas.
- [ ] I/we have made any necessary changes to documentation. -- will do this in a separate PR
- [x] I/we have added tests that cover new code.
- [x] I/we have run tests and they pass locally with the changes.
- [x] I/we have run `go fmt ./...` and `golangci-lint run`.

### CODE REVIEW

File: internal/api/client/interactionrequests/authorize.go [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+
+	"github.com/gin-gonic/gin"
+	apiutil "github.com/superseriousbusiness/gotosocial/internal/api/util"
+	"github.com/superseriousbusiness/gotosocial/internal/gtserror"
+	"github.com/superseriousbusiness/gotosocial/internal/oauth"
+)
+
+// InteractionRequestAuthorizePOSTHandler swagger:operation POST /api/v1/interaction_requests/{id}/authorize authorizeInteractionRequest
+//
+// Accept/authorize/approve an interaction request with the given ID.
```
Comment (tsmethurst) [inline internal/api/client/interactionrequests/authorize.go:31]: Used the same verbiage here (authorize) as we use for follow requests, for consistency.

File: testrig/teststructs.go [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package testrig
```
Comment (tsmethurst) [inline testrig/teststructs.go:18]: Moved this stuff to a separate package to avoid repetition + make our lives a bit easier when testing things that rely on workers for side effects.

File: internal/processing/workers/workers_test.go [hunk 3]
```diff
@@ -132,63 +114,3 @@ func (suite *WorkersTestSuite) openStreams(ctx context.Context, processor *proce
 
 	return streams
 }
-
-func (suite *WorkersTestSuite) SetupTestStructs() *TestStructs {
```
Comment (tsmethurst) [inline internal/processing/workers/workers_test.go:136]: Moved this to the testrig as it's useful elsewhere.

File: internal/db/bundb/timeline.go [hunk 4]
```diff
@@ -137,6 +124,22 @@ func (t *timelineDB) GetHomeTimeline(ctx context.Context, accountID string, maxI
 		bun.In(targetAccountIDs),
 	)
 
+	// Only include statuses that aren't pending approval.
+	q = q.Where("NOT ? = ?", bun.Ident("status.pending_approval"), true)
```
Comment (tsmethurst) [inline internal/db/bundb/timeline.go:128]: Added this clause here and elsewhere in this package so that timelines don't have pending statuses put in them. Let me know if you'd prefer this in a separate PR.

File: internal/db/bundb/instance.go [hunk 5]
```diff
@@ -76,6 +76,9 @@ func (i *instanceDB) CountInstanceStatuses(ctx context.Context, domain string) (
 			Where("? = ?", bun.Ident("account.domain"), domain)
 	}
 
+	// Ignore statuses that are currently pending approval.
+	q = q.Where("NOT ? = ?", bun.Ident("status.pending_approval"), true)
```
Comment (tsmethurst) [inline internal/db/bundb/instance.go:80]: See https://github.com/superseriousbusiness/gotosocial/pull/3215/files#r1723451114, another thing I can put in a separate PR if you prefer.

File: internal/db/bundb/account.go [hunk 6]
```diff
@@ -1285,34 +1285,40 @@ func (a *accountDB) RegenerateAccountStats(ctx context.Context, account *gtsmode
 	if err := a.db.RunInTx(ctx, nil, func(ctx context.Context, tx bun.Tx) error {
 		var err error
 
-		// Scan database for account statuses.
+		// Scan database for account statuses, ignoring
+		// statuses that are currently pending approval.
 		statusesCount, err := tx.NewSelect().
-			Table("statuses").
-			Where("? = ?", bun.Ident("account_id"), account.ID).
+			TableExpr("? AS ?", bun.Ident("statuses"), bun.Ident("status")).
```
Comment (tsmethurst) [inline internal/db/bundb/account.go:1291]: See https://github.com/superseriousbusiness/gotosocial/pull/3215/files#r1723451114, another thing I can put in a separate PR if you prefer.

File: internal/db/bundb/migrations/20240716151327_interaction_policy.go [hunk 7]
```diff
@@ -1,71 +0,0 @@
...
-package migrations
-
-import (
-	"context"
-
-	gtsmodel "github.com/superseriousbusiness/gotosocial/internal/gtsmodel"
-	"github.com/uptrace/bun"
-)
-
-func init() {
```
Comment (tsmethurst) [inline internal/db/bundb/migrations/20240716151327_interaction_policy.go:27]: Nuked this migration since the one i added replaces everything in it.

File: internal/db/bundb/migrations/20240809134448_interaction_requests_client_api.go [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+				IfNotExists().
+				Exec(ctx); err != nil {
+				return err
+			}
+
+			for idx, col := range map[string]string{
+				"interaction_rejections_status_id_idx":              "status_id",
+				"interaction_rejections_account_id_idx":             "account_id",
+				"interaction_rejections_interacting_account_id_idx": "interacting_account_id",
+				"interaction_rejections_interaction_uri_idx":        "interaction_uri",
```
Comment (tsmethurst) [inline internal/db/bundb/migrations/20240809134448_interaction_requests_client_api.go:181]: Please note that unlike with the other interaction models,  `*gtsmodel.InteractionRejection.InteractionURI` is NOT unique, and so an index needs to be created for it. It's possible that more than one person might want to reject an interaction.

File: internal/db/bundb/interaction_test.go [hunk 9]
```diff
@@ -0,0 +1,264 @@
...
+	"github.com/superseriousbusiness/gotosocial/internal/paging"
+	"github.com/superseriousbusiness/gotosocial/internal/typeutils"
+	"github.com/superseriousbusiness/gotosocial/internal/util"
+)
+
+type InteractionTestSuite struct {
+	BunDBStandardTestSuite
+}
+
+func (suite *InteractionTestSuite) markIntsPending(
```
Comment (NyaaaWhatsUpDoc) [inline internal/db/bundb/interaction_test.go:38]: i think it might read a bit better if this was just "markPending". on first read i was wondering where the integers were :p

File: internal/processing/workers/util.go [hunk 10]
```diff
@@ -488,128 +487,143 @@ func (u *utils) decrementFollowRequestsCount(
 	return nil
 }
 
-// approveFave stores + returns an
-// interactionApproval for a fave.
-func (u *utils) approveFave(
+// pendFave stores an interaction request for the given
+// fave, and notifies the interactee of a pending fave.
+func (u *utils) pendFave(
```
Comment (NyaaaWhatsUpDoc) [inline internal/processing/workers/util.go:492]: i think these might read a bit better if they were `setPendingFave()`, `setPendingReply()` etc. what do you think?

File: internal/processing/workers/fromclientapi.go [hunk 11]
```diff
@@ -1050,16 +1053,206 @@ func (p *clientAPI) RejectUser(ctx context.Context, cMsg *messages.FromClientAPI
...
+	if !ok {
+		return gtserror.Newf("%T not parseable as *gtsmodel.InteractionRejection", cMsg.GTSModel)
+	}
+
+	// At this point the InteractionRejection should already
+	// be in the database, we just need to do side effects.
+
+	// Send out the rejection as Reject.
+	if err := p.federate.RejectInteraction(ctx, rejection); err != nil {
+		return gtserror.Newf("error federating rejection of reply: %w", err)
```
Comment (NyaaaWhatsUpDoc) [inline internal/processing/workers/fromclientapi.go:1197]: i would instead just make this a `log.Error(...)` to ensure we always end up reaching the stages below to wipe the status.

File: internal/processing/workers/fromclientapi.go [hunk 12]
```diff
@@ -1050,16 +1053,206 @@ func (p *clientAPI) RejectUser(ctx context.Context, cMsg *messages.FromClientAPI
...
+	if !ok {
+		return gtserror.Newf("%T not parseable as *gtsmodel.InteractionRejection", cMsg.GTSModel)
+	}
+
+	// At this point the InteractionRejection should already
+	// be in the database, we just need to do side effects.
+
+	// Send out the rejection as Reject.
+	if err := p.federate.RejectInteraction(ctx, rejection); err != nil {
+		return gtserror.Newf("error federating rejection of like: %w", err)
```
Comment (NyaaaWhatsUpDoc) [inline internal/processing/workers/fromclientapi.go:1160]: i would instead just make this a `log.Error(...)` to ensure we always end up reaching the stages below to delete the fave.

File: internal/processing/workers/fromclientapi.go [hunk 13]
```diff
@@ -1050,16 +1053,206 @@ func (p *clientAPI) RejectUser(ctx context.Context, cMsg *messages.FromClientAPI
...
+	if err := p.federate.RejectInteraction(ctx, rejection); err != nil {
+		return gtserror.Newf("error federating rejection of reply: %w", err)
+	}
+
+	// Get the rejected status.
+	status, err := p.state.DB.GetStatusByURI(
+		gtscontext.SetBarebones(ctx),
+		rejection.InteractionURI,
+	)
+	if err != nil && !errors.Is(err, db.ErrNoEntries) {
```
Comment (NyaaaWhatsUpDoc) [inline internal/processing/workers/fromclientapi.go:1205]: i think it's probably worth *not* checking for errnoentries here. in this case if the status is already deleted it actually is an error we want to return with.

File: internal/processing/workers/fromclientapi.go [hunk 14]
```diff
@@ -1050,16 +1053,206 @@ func (p *clientAPI) RejectUser(ctx context.Context, cMsg *messages.FromClientAPI
...
+		return gtserror.Newf("error federating rejection of announce: %w", err)
+	}
+
+	// Get the rejected boost.
+	boost, err := p.state.DB.GetStatusByURI(
+		gtscontext.SetBarebones(ctx),
+		rejection.InteractionURI,
+	)
+	if err != nil && !errors.Is(err, db.ErrNoEntries) {
+		return gtserror.Newf("db error getting rejected announce: %w", err)
```
Comment (NyaaaWhatsUpDoc) [inline internal/processing/workers/fromclientapi.go:1243]: i think it's probably worth *not* checking for errnoentries here. in this case if the status is already deleted it actually is an error we want to return with.

Comment (NyaaaWhatsUpDoc) [reply internal/processing/workers/fromclientapi.go:1243]: (also simpler code)

File: internal/processing/workers/fromclientapi.go (hunk 13 above)
Comment (NyaaaWhatsUpDoc) [reply internal/processing/workers/fromclientapi.go:1205]: (also simpler code)

File: internal/processing/workers/fromclientapi.go [hunk 15]
```diff
@@ -1050,16 +1053,206 @@ func (p *clientAPI) RejectUser(ctx context.Context, cMsg *messages.FromClientAPI
...
+	if err := p.federate.RejectInteraction(ctx, rejection); err != nil {
+		return gtserror.Newf("error federating rejection of like: %w", err)
+	}
+
+	// Get the rejected fave.
+	fave, err := p.state.DB.GetStatusFaveByURI(
+		gtscontext.SetBarebones(ctx),
+		rejection.InteractionURI,
+	)
+	if err != nil && !errors.Is(err, db.ErrNoEntries) {
```
Comment (NyaaaWhatsUpDoc) [inline internal/processing/workers/fromclientapi.go:1168]: i think it's probably worth not checking for errnoentries here. in this case if the status is already deleted it actually is an error we want to return with. (also simpler code)

File: internal/processing/workers/fromclientapi.go [hunk 16]
```diff
@@ -1050,16 +1053,206 @@ func (p *clientAPI) RejectUser(ctx context.Context, cMsg *messages.FromClientAPI
...
+	)
+
+	// Update stats for the reply author account.
+	if err := p.utils.incrementStatusesCount(ctx, interactingAcct, reply); err != nil {
+		log.Errorf(ctx, "error updating account stats: %v", err)
+	}
+
+	// Interaction counts changed on the replied status;
+	// uncache the prepared version from all timelines.
+	p.surface.invalidateStatusFromTimelines(ctx, reply.InReplyToID)
```
Comment (NyaaaWhatsUpDoc) [inline internal/processing/workers/fromclientapi.go:1096]: does it matter that this `invalidateStatusFromTimelines()` call here relative to the other similar calls in both `AcceptAnnounce()` and `AcceptReply()` are in different locations?

Comment (NyaaaWhatsUpDoc) [reply internal/processing/workers/fromclientapi.go:1096]: and it's probably worth standardising the behaviour in all these functions e.g. invalidate steps always go *after* <some particular logic>. if you get what i mean :)

File: internal/processing/workers/fromfediapi.go [hunk 17]
```diff
@@ -228,24 +230,24 @@ func (p *fediAPI) CreateStatus(ctx context.Context, fMsg *messages.FromFediAPI)
 		return nil
 	}
 
+	// We're going to possibly modify
+	// the status further, lock on its URI.
+	unlock := p.state.ProcessingLocks.Lock(status.URI)
```
[the remark this answers is not in the document]
Comment (NyaaaWhatsUpDoc) [reply internal/processing/workers/fromfediapi.go:235]: also given we're doing these in the `pend___()` functions i think we're doubling up on locks here

File: internal/processing/interactionrequests/reject.go [hunk 18]
```diff
@@ -0,0 +1,209 @@
...
+	"github.com/superseriousbusiness/gotosocial/internal/gtsmodel"
+	"github.com/superseriousbusiness/gotosocial/internal/messages"
+)
+
+// Reject rejects an interaction request with the given ID,
+// on behalf of the requester (whose post it must target).
+func (p *Processor) Reject(
+	ctx context.Context,
+	requester *gtsmodel.Account,
+	intReqID string,
```
Comment (NyaaaWhatsUpDoc) [inline internal/processing/interactionrequests/reject.go:37]: i think this would be a bit easier to read if it was just `requestID`, or `id`. `intReqID` reads like there should be an integer in there :p

File: internal/processing/interactionrequests/accept.go [hunk 19]
```diff
@@ -0,0 +1,209 @@
...
+
+	return apiApproval, nil
+}
+
+// Package-internal convenience function to accept a like.
+func (p *Processor) acceptLike(
+	ctx context.Context,
+	req *gtsmodel.InteractionRequest,
+) (*gtsmodel.InteractionApproval, gtserror.WithCode) {
+	if req.Like == nil {
```
Comment (NyaaaWhatsUpDoc) [inline internal/processing/interactionrequests/accept.go:117]: i don't think we need to check for these here. if they were nil but the ID for the object was there as it should be, then the initial GetInteractionRequest() to the database should have returned error in the final stage populating it. Adding these nil checks here makes it seem like this is an (at least somewhat) expected situation. Absolute worst-case it'll just panic on attempting to deref ptr below.

File: internal/processing/interactionrequests/accept.go [hunk 20]
```diff
@@ -0,0 +1,209 @@
...
+
+	return approval, nil
+}
+
+// Package-internal convenience function to accept a reply.
+func (p *Processor) acceptReply(
+	ctx context.Context,
+	req *gtsmodel.InteractionRequest,
+) (*gtsmodel.InteractionApproval, gtserror.WithCode) {
+	if req.Reply == nil {
```
Comment (NyaaaWhatsUpDoc) [inline internal/processing/interactionrequests/accept.go:150]: same as above re: nil check.

File: internal/processing/interactionrequests/accept.go [hunk 21]
```diff
@@ -0,0 +1,209 @@
...
+
+	return approval, nil
+}
+
+// Package-internal convenience function to accept an announce.
+func (p *Processor) acceptAnnounce(
+	ctx context.Context,
+	req *gtsmodel.InteractionRequest,
+) (*gtsmodel.InteractionApproval, gtserror.WithCode) {
+	if req.Announce == nil {
```
Comment (NyaaaWhatsUpDoc) [inline internal/processing/interactionrequests/accept.go:183]: same as above re: nil check.

File: internal/processing/interactionrequests/reject.go [hunk 22]
```diff
@@ -0,0 +1,209 @@
...
+
+	return apiRejection, nil
+}
+
+// Package-internal convenience function to reject a like.
+func (p *Processor) rejectLike(
+	ctx context.Context,
+	req *gtsmodel.InteractionRequest,
+) (*gtsmodel.InteractionRejection, gtserror.WithCode) {
+	if req.Like == nil {
```
Comment (NyaaaWhatsUpDoc) [inline internal/processing/interactionrequests/reject.go:117]: same as above re: nil check.

File: internal/processing/interactionrequests/reject.go [hunk 23]
```diff
@@ -0,0 +1,209 @@
...
+
+	return rejection, nil
+}
+
+// Package-internal convenience function to reject a reply.
+func (p *Processor) rejectReply(
+	ctx context.Context,
+	req *gtsmodel.InteractionRequest,
+) (*gtsmodel.InteractionRejection, gtserror.WithCode) {
+	if req.Reply == nil {
```
Comment (NyaaaWhatsUpDoc) [inline internal/processing/interactionrequests/reject.go:150]: same as above re: nil check

File: internal/processing/interactionrequests/reject.go [hunk 24]
```diff
@@ -0,0 +1,209 @@
...
+
+	return rejection, nil
+}
+
+// Package-internal convenience function to reject an announce.
+func (p *Processor) rejectAnnounce(
+	ctx context.Context,
+	req *gtsmodel.InteractionRequest,
+) (*gtsmodel.InteractionRejection, gtserror.WithCode) {
+	if req.Announce == nil {
```
Comment (NyaaaWhatsUpDoc) [inline internal/processing/interactionrequests/reject.go:183]: same as above re: nil check.

File: internal/processing/interactionrequests/accept.go (hunk 19 above)
Comment (NyaaaWhatsUpDoc) [reply internal/processing/interactionrequests/accept.go:117]: also if that does happen, it could potentially be a serious issue. a not-found is likely to end up getting masked and we won't end up dealing with it, whereas a panic and unexpected 500 definitely does get our attention :p

File: internal/api/client/interactionrequests/authorize.go (hunk 1 above)
Comment (NyaaaWhatsUpDoc) [reply internal/api/client/interactionrequests/authorize.go:31]: what does the authorize here mean then? i'm not actually sure what the authorize means in that context either.

File: internal/api/client/interactionrequests/getpage.go [hunk 25]
```diff
@@ -0,0 +1,211 @@
...
+	includeBoosts, errWithCode := apiutil.ParseInteractionReblogs(
+		c.Query(apiutil.InteractionReblogsKey), true,
+	)
+	if errWithCode != nil {
+		apiutil.ErrorHandler(c, errWithCode, m.processor.InstanceGetV1)
+		return
+	}
+
+	if !includeLikes && !includeReplies && !includeBoosts {
+		const text = "at least one of favourites, replies, or boosts must be true"
```
Comment (NyaaaWhatsUpDoc) [inline internal/api/client/interactionrequests/getpage.go:176]: i think the api response may make a bit more sense if this was `must be set`

File: internal/db/bundb/account.go (hunk 6 above)
Comment (NyaaaWhatsUpDoc) [reply internal/db/bundb/account.go:1291]: nah 'tis all good :+1:

File: internal/api/model/interaction.go [hunk 26]
```diff
@@ -0,0 +1,82 @@
...
+	// The type of interaction that this interaction request pertains to.
+	//
+	//	`favourite` - Someone favourited a status.
+	//	`reply` - Someone replied to a status.
+	//	`reblog` - Someone reblogged / boosted a status.
+	Type string `json:"type"`
+	// The timestamp of the interaction request (ISO 8601 Datetime)
+	CreatedAt string `json:"created_at"`
+	// The account that performed the interaction.
+	InteractingAccount *Account `json:"interacting_account"`
```
Comment (NyaaaWhatsUpDoc) [inline internal/api/model/interaction.go:35]: very subjective one here, but i think these would make more sense as just `account` and `status` respectively (same goes for the structs below). it fits a bit better with the `reply` json key imo

Comment (NyaaaWhatsUpDoc) [reply internal/api/model/interaction.go:35]: (also makes json a little easier to read with keys that aren't too wordy)

File: internal/db/bundb/account.go [hunk 27]
```diff
@@ -1285,34 +1285,40 @@ func (a *accountDB) RegenerateAccountStats(ctx context.Context, account *gtsmode
...
+		// Scan database for pinned statuses, ignoring
+		// statuses that are currently pending approval.
 		statusesPinnedCount, err := tx.NewSelect().
-			Table("statuses").
-			Where("? = ?", bun.Ident("account_id"), account.ID).
-			Where("? IS NOT NULL", bun.Ident("pinned_at")).
+			TableExpr("? AS ?", bun.Ident("statuses"), bun.Ident("status")).
+			Where("? = ?", bun.Ident("status.account_id"), account.ID).
+			Where("? IS NOT NULL", bun.Ident("status.pinned_at")).
+			Where("NOT ? = ?", bun.Ident("status.pending_approval"), true).
```
Comment (NyaaaWhatsUpDoc) [inline internal/db/bundb/account.go:1306]: is this actually possible for a pinned status?

File: internal/api/client/interactionrequests/authorize.go (hunk 1 above)
Comment (tsmethurst) [reply internal/api/client/interactionrequests/authorize.go:31]: Authorize = accept, basically. Not sure why "authorize/reject" is used in the Masto follow requests API instead of just "accept/reject", but that's life!

File: internal/db/bundb/migrations/20240716151327_interaction_policy.go (hunk 7 above)
Comment (tsmethurst) [reply internal/db/bundb/migrations/20240716151327_interaction_policy.go:27]: Oh actually you know, I'm not sure. I'll just make it an empty migration then, just in case.

File: internal/processing/workers/util.go (hunk 10 above)
Comment (tsmethurst) [reply internal/processing/workers/util.go:492]: I'm not sure... At this point the fave (or whatever) is already marked as pending, so nothing is being set on that fave itself, this is more about handling the side effects of a pending fave (ie., putting an interaction request in the db), notifying the target. Maybe handlePendingFave is better?

File: internal/db/bundb/interaction.go [hunk 28]
```diff
@@ -128,22 +142,445 @@ func (r *interactionDB) PopulateInteractionApproval(ctx context.Context, approva
 		}
 	}
 
+	switch approval.InteractionType {
+
+	case gtsmodel.InteractionLike:
+		approval.Like, err = i.state.DB.GetStatusFaveByURI(ctx, approval.InteractionURI)
+		if err != nil {
+			errs.Appendf("error populating interactionApproval Like")
```
Comment (NyaaaWhatsUpDoc) [inline internal/db/bundb/interaction.go:150]: i assume you meant to wrap the error here too :p

File: internal/db/bundb/interaction_test.go (hunk 9 above)
Comment (tsmethurst) [reply internal/db/bundb/interaction_test.go:38]: Got ya, yeah that's a bit weird, I'll change it.

File: internal/db/bundb/account.go (hunk 27 above)
Comment (tsmethurst) [reply internal/db/bundb/account.go:1306]: Yeah, I think you could create a reply to someone, and then try to pin your reply. Or something daft like that. I suppose we could just make it so that you can't pin pending statuses, instead.

File: internal/api/model/interaction.go (hunk 26 above)
Comment (tsmethurst) [reply internal/api/model/interaction.go:35]: Mmm that's what I originally had but it felt a bit weird from a client perspective so I changed it to this. I'll change it back though (it's more consistent with the Masto API anyways).

File: internal/db/bundb/interaction.go (hunk 28 above)
Comment (tsmethurst) [reply internal/db/bundb/interaction.go]: Yes, thanks, fixed! https://github.com/superseriousbusiness/gotosocial/pull/3215/commits/9fb07cb443cf55e0dd4eab287f6ddfcc7b4cff8b

File: internal/api/client/interactionrequests/getpage.go (hunk 25 above)
Comment (tsmethurst) [reply internal/api/client/interactionrequests/getpage.go:176]: I'd rather leave it as is, since "must be set" would also suggest you could set them all, but to false.

File: internal/processing/interactionrequests/accept.go (hunk 20 above)
Comment (tsmethurst) [reply internal/processing/interactionrequests/accept.go:168]: Resolving + using above comment for discussion.

File: internal/processing/interactionrequests/accept.go (hunk 21 above)
Comment (tsmethurst) [reply internal/processing/interactionrequests/accept.go]: Resolving + using above comment for discussion.

File: internal/processing/interactionrequests/reject.go (hunk 22 above)
Comment (tsmethurst) [reply internal/processing/interactionrequests/reject.go]: Resolving + using above comment for discussion.

File: internal/processing/interactionrequests/reject.go (hunk 23 above)
Comment (tsmethurst) [reply internal/processing/interactionrequests/reject.go]: Resolving + using above comment for discussion.

File: internal/processing/interactionrequests/reject.go (hunk 24 above)
Comment (tsmethurst) [reply internal/processing/interactionrequests/reject.go]: Resolving + using above comment for discussion.

File: internal/processing/workers/fromclientapi.go (hunk 15 above)
Comment (tsmethurst) [reply internal/processing/workers/fromclientapi.go]: https://github.com/superseriousbusiness/gotosocial/pull/3215/commits/73a7845919f0d0039698c2efcd9cb0bda360cff4

File: internal/processing/workers/fromclientapi.go (hunk 14 above)
Comment (tsmethurst) [reply internal/processing/workers/fromclientapi.go:1306]: https://github.com/superseriousbusiness/gotosocial/pull/3215/commits/73a7845919f0d0039698c2efcd9cb0bda360cff4

File: internal/processing/workers/fromclientapi.go (hunk 13 above)
Comment (tsmethurst) [reply internal/processing/workers/fromclientapi.go]: https://github.com/superseriousbusiness/gotosocial/pull/3215/commits/73a7845919f0d0039698c2efcd9cb0bda360cff4

File: internal/processing/interactionrequests/accept.go (hunk 19 above)
Comment (tsmethurst) [reply internal/processing/interactionrequests/accept.go:127]: I updated the db functions to be OK with NotFound errors when trying to fetch a Like, Reply, or Announce to populate an interaction request. Especially now that rejections are using the same model as accepts and requests, there's a fair chance that if someone tries to fetch an interaction request from the db, and it's for something that's been rejected, the Like, Reply, or Announce won't be in the database anymore, so it is a semi-expected error. Given that, I think it's sensible to have nil checking here. I've changed it now so that an error not found is returned if the Like (or whatever) can't be found, which should be warning enough I think.

File: internal/processing/interactionrequests/reject.go (hunk 18 above)
Comment (tsmethurst) [reply internal/processing/interactionrequests/reject.go]: You're right! changed: https://github.com/superseriousbusiness/gotosocial/pull/3215/commits/5a127b64c4384793b12148c17e1d721acf8e3eff

File: internal/processing/workers/util.go (hunk 10 above)
Comment (tsmethurst) [reply internal/processing/workers/util.go]: I changed them to "requestFave" etc. I think that makes a bit more sense!

File: internal/processing/workers/fromclientapi.go (hunk 12 above)
Comment (tsmethurst) [reply internal/processing/workers/fromclientapi.go]: p.federate.RejectInteraction

File: internal/processing/workers/fromclientapi.go (hunk 11 above)
Comment (tsmethurst) [reply internal/processing/workers/fromclientapi.go]: https://github.com/superseriousbusiness/gotosocial/pull/3215/commits/1542bfcc99a0e010749ad2e8527d79819740ba07

File: internal/api/model/interaction.go (hunk 26 above)
Comment (tsmethurst) [reply internal/api/model/interaction.go]: Done :)

File: internal/db/bundb/migrations/20240716151327_interaction_policy.go (hunk 7 above)
Comment (tsmethurst) [reply internal/db/bundb/migrations/20240716151327_interaction_policy.go:27]: Done!

File: docs/api/swagger.yaml [hunk 29]
```diff
@@ -7521,6 +7562,177 @@ paths:
...
+                - application/json
+            responses:
+                "200":
+                    description: The now-rejected interaction request.
+                    schema:
+                        $ref: '#/definitions/interactionRequest'
+                "400":
+                    description: bad request
+                "401":
+                    description: unrejectd
```
Comment (VyrCossont) [inline docs/api/swagger.yaml:7723]: Typo, probably should be "unauthorized"

Comment (tsmethurst) [reply docs/api/swagger.yaml:7723]: Ah thanks, search+replace error :')

Comment (tsmethurst) [reply docs/api/swagger.yaml]: https://github.com/superseriousbusiness/gotosocial/pull/3215/commits/7f9f8c6a11c19a49c6e6574ff9c3c6827ccb5b9a

File: internal/processing/workers/fromfediapi.go (hunk 17 above)
[the remark this answers is not in the document]
Comment (tsmethurst) [reply internal/processing/workers/fromfediapi.go]: Right! Changed it :') https://github.com/superseriousbusiness/gotosocial/pull/3215/commits/30dddd22a868b5cc0c8f4dcaf00536a61a3c9b75

File: internal/processing/workers/fromclientapi.go (hunk 16 above)
Comment (tsmethurst) [reply internal/processing/workers/fromclientapi.go]: I get ya! Changed it here https://github.com/superseriousbusiness/gotosocial/pull/3215/commits/e62f6d5aef37430621babfc97bce172a9c1eeb35. I think functionally it doesn't matter that much, but it is neater to have it always be the last step indeed.
